### DISCUSSION TITLE
A way to add host entries to Mission Portal through an API?

### DISCUSSION THREAD

**Discussion by craigcomstock** [cfengine/core#5443] [Q&A] [answered]
I want to scrape the connected hosts to my WiFi router and ensure they each have a created host entry in Mission Portal.

I see a Host API but no way to create entries there: https://docs.cfengine.com/docs/3.21/api-enterprise-api-ref-host.html

Any ideas? I am running the hub in a docker container on an toradex/torizon dev board FWIW, works fine. 👍 

I am trying to integrate with an OpenWRT wifi router which I can ssh to to scrape the information. Ideally I'd like to install an Enterprise client there but haven't gotten around to building that yet. :)

**Comment by craigcomstock** [comment]
It would be cool if there was a way in policy to create the host on the policy server. That way I could parse the dhcp lease information from openwrt and add hosts as needed via policy/script. Ideally I'd lookup the hosts by mac address and update/create based on that key.

**Comment by olehermanse** [comment]
@craigcomstock There is no good / supported way to add hosts to Mission Portal without running CFEngine on them. Do you want to add them to get an overview of which hosts still need to have CFEngine installed, or do you want to keep these "fake" host entries forever?

**Reply by craigcomstock** [reply]
I do want to keep these fake entries forever. Like an inventory of technology items in my environment. I was thinking too of adding various connected devices even which can't have agents like bluetooth headphones which report battery through BT, keyboards, USB devices, the lot.

I did manage to get something working as a shell script which looks at a dhcp.leases file and runs various psql commands to manipulate the database. I understand it is a "hack" and may have many odd repercussions but I'm ok with that in a development/hobby environment. :) 

I was also considering adding some sort of linking between hosts to qualify their connection, like "bluetooth client", "docker instance", "vagrant instance", "ssh proxy" (like for my openwrt router). And really at that point a "host" is really more like a "device" or some object with a micro processor or micro controller... something that runs code and connects in some way.

I noticed that bluetooth on macos reports firmware versions so I thought I might check those with CVEs and such and/or check for updates.

the command
```
system_profiler SPBluetoothDataType -json
```

and one device out of that output :) I imagine I could have a "host" and track firmware version and battery level and then push that up to a compliance report for firmware version up-to-date and a report/alert for battery is low.

```
        {
          "UE BOOM" : {
            "device_address" : "00:0D:44:DF:E2:EB",
            "device_firmwareVersion" : "255.15.15",
            "device_minorType" : "Headphones",
            "device_productID" : "0xFFFF",
            "device_vendorID" : "0x000A"
          }
        }
```

**Reply by olehermanse** [reply, marked as answer]
I see. This is not a highly requested feature, so likely not something we will add more support for in the near future. Cool that you got something working :)

**Reply by craigcomstock** [reply]
Thanks! Look forward to working on this more as I can. Maybe submit some modules to https://build.cfengine.com/ 

Cheers.
